### PR TITLE
chore: Upgrade to syn v2.x and svgbob v0.7

### PR DESCRIPTION
**Pull request by jetuk**
Upgrade to syn v2.x and svgbob v0.7. Then address the resulting Clippy issues.

### CODE REVIEW

File: rust-toolchain.toml
Comment (yvt) [inline rust-toolchain.toml]:
```suggestion
channel = "1.66"
```

File: Cargo.toml
Comment (yvt) [inline Cargo.toml]:
```suggestion
base64 = ">= 0.21, < 0.23"
```

File: Cargo.toml
Comment (yvt) [inline Cargo.toml]:
```suggestion
syn = "2"
```

File: Cargo.toml
Comment (yvt) [inline Cargo.toml]: The difference between `^0.1` and `^0.2` is insignificant for us (afaik), but `svgbob ^0.7` still uses `^0.1`, so forcing the bump to `^0.2` would increase the build time.

```suggestion
unicode-width = ">= 0.1, < 0.3"
```

**Review by yvt — CHANGES_REQUESTED** [review]
Sorry for the belated response! This looks good aside from a few minor issues.

**Comment by Sympatron** [comment]
A release after this is merged would be nice. :)

**Comment by jetuk** [comment]
Thanks for the review. I've made the updates in the latest commit and updated the Rust version in the README.

**Comment by jetuk** [comment]
>  error: package `bumpalo v3.16.0` cannot be built because it requires rustc 1.73.0 or newer, while the currently active rustc version is 1.66.1

This is failing on the above. Would you prefer to bump the MSRV or downgrade `bumpalo`?

**Comment by yvt** [comment]
Could you add `Cargo.lock` to the repo (removing it from `.gitignore`)?

Following the [default behavior](https://doc.rust-lang.org/cargo/faq.html#why-have-cargolock-in-version-control) of an older version of Cargo, `Cargo.lock` is currently excluded from the repository, which results in Cargo picking the latest dependencies on every CI run. Not only does this hurt build determinism, but it can also spontaneously break CI builds because Cargo can pick a crate version (`bumpalo v3.16.0` in this case) that is incompatible with the current compiler (the MSRV-aware resolver is still unstable). I think it's time we switched to the new behavior.

**Comment by jetuk** [comment]
I've tried this and it requires going back to 3.14.0 which then has some build failures elsewhere in the dependency tree. I think it will be easier to bump MSRV to 1.73.0 to align with `bumpalo`. 

Also, I think the actions (derived from action-rs) are also unmaintained. I can update those to dtolnay actions which seem to be quite widely used now if you like?

**Comment by yvt** [comment]
I see, let's bump to 1.73 then.

Upgrading the CI actions would be nice, but please do this in a separate PR.

File: README.md
Comment (yvt) [inline README.md]:
```suggestion
*Requires Rust version 1.73 or later or equivalent nightly builds.*
```

**Review by yvt — CHANGES_REQUESTED** [review]

**Review by yvt — APPROVED** [review]

**Comment by yvt** [comment]
Thanks!

**Comment by Sympatron** [comment]
After this is merged now a release would be welcome :)

**Comment by ROMemories** [comment]
Thanks for updating these dependencies! Could you cut a new release? `svgbobdoc` is one of the last users of `syn` v1 in our tree.

**Comment by yvt** [comment]
@ROMemories Do you have any opinions on whether it should be a major version bump? On the one hand, bumping svgbob can break existing diagrams. But on the other hand, breaking diagrams could be insignificant compared to dependency hell.

**Comment by Sympatron** [comment]
If svgbobdoc does not expose anything syn specific I don't think a major version bump is necessary. But if in doubt, you can use [cargo-semver-checks](https://github.com/obi1kenobi/cargo-semver-checks) to be sure.

**Comment by ROMemories** [comment]
> Do you have any opinions on whether it should be a major version bump? On the one hand, bumping svgbob can break existing diagrams.

@yvt If it *can* break diagrams I do think there should be an incompatible version bump. However, looking at the [changelog of `svgbob`](https://github.com/ivanceras/svgbob/blob/master/Changelog.md#070), it seems to me that the only breaking change was about how the styles were handled, which I suppose should be entirely contained by `svgbobdoc`?

---

To be clear, we're not using `svgbobdoc` directly, it is only in our tree through `rlsf` ("we" meaning [ariel-os/ariel-os](https://github.com/ariel-os/ariel-os)).

**Comment by Sympatron** [comment]
I totally missed that this PR updates `svgbob` as well... In that case it definitely has to be a major version bump.

What do you think about splitting this PR? Release 0.3.1 that just updates `syn` and then 0.4.0 with the update of `svgbob`.
That way crates like `rlsf` could profit immediately. Since rlsf wasn't updated in 3 years, I would rather not wait for them to do a release before downstream crates can get rid of `syn 1`…

I have a fork that just has the semver compatible parts of this PR [here](https://github.com/Sympatron/svgbobdoc/tree/syn-update), but I can't open a PR, because you can only open a PR against branches and not tags. You would have to create a branch at tag `0.3.0` for me to be able to open a PR.

**Comment by yvt** [comment]
> If it _can_ break diagrams I do think there should be an incompatible version bump. However, looking at the [changelog of `svgbob`](https://github.com/ivanceras/svgbob/blob/master/Changelog.md#070), it seems to me that the only breaking change was about how the styles were handled, which I suppose should be entirely contained by `svgbobdoc`?

Technically, any changes to rendering (e.g., supporting three-quarter circles) can introduce breakage in existing diagrams because there are no input patterns reserved for future extension. `svgbob`'s semver scheme doesn't use this strict definition of breakage, as it can be seen that it has introduced rendering-changing changes in minor updates in the past (examples: [0.6.1](https://github.com/ivanceras/svgbob/commit/dfb72d701bb5bd0e59e8bee4daa86d6861d3ac46), [0.7.6](https://github.com/ivanceras/svgbob/commit/2f4f8f578a4fee2dbb398c3778229330829af88b)). This means if we followed this definition to the letter while keeping `svgbob` up-to-date, we'd have to pin the version of `svgbob` and do a major version bump even on a minor version bump of `svgbob`, which would negatively affect maintenance (lots of branches to maintain) and users' dependency graphs (multiple copies of `svgbob[doc]`).

For this reason, I'm somewhat inclined to tolerate some rendering-breaking changes as long as they don't break trivial diagrams, such as those made up of just hollow rectangles, perpendicular lines, and texts, meaning that the update to `svgbob` v0.7 is non-breaking.

**Pull request merged**
